modules/ocf_www/files/build-vhosts

```diff
@@ -14,190 +11,350 @@ from pathlib import Path
 
 import jinja2
 from ocflib.account.search import user_is_sorried
-from ocflib.account.utils import web_dir
 from ocflib.misc.mail import email_for_user
+from ocflib.vhost.application import get_app_vhosts
 from ocflib.vhost.web import get_vhosts
 
+
+APACHE_SITE_CONFIG = '/etc/apache2/ocf-vhost.conf'
+NGINX_SITE_CONFIG = '/etc/nginx/sites-enabled/virtual'
+
+LETS_ENCRYPT_SSL = Path('/services/http/ssl')
+SSL_BUNDLE = Path('/etc/ssl/apphost')
+
+APP_DIR = Path('/srv/apps')
+WEB_DIR = Path('/services/http/users')
+
+
 jinja_env = jinja2.Environment(
     loader=jinja2.FileSystemLoader((
-        os.path.abspath(os.path.dirname(__file__)),
-        '/opt/share/',
+        os.path.abspath('/opt/share'),
     )),
 )
 
 
-LETS_ENCRYPT_SSL = Path('/services/http/ssl')
+class SSL(namedtuple('SSL', ('fqdn',))):
+    """Tells where to find SSL files for a given domain."""
+    @property
+    def is_valid(self):
+        return os.path.exists(self.cert)
+
+    @property
+    def key(self):
+        return '/etc/ssl/lets-encrypt/le-vhost.key'
+
+    @property
+    def cert(self):
+        return str(LETS_ENCRYPT_SSL / (self.fqdn + '.crt'))
+
+    @property
+    def chain(self):
+        return '/etc/ssl/certs/lets-encrypt.crt'
+
+    @property
+    def bundle(self):
+        return str(SSL_BUNDLE / (self.fqdn + '.bundle'))
 
 
 class VirtualHost(namedtuple('VirtualHost', (
-        'fqdn',
-        'user',
-        'relative_docroot',
-        'redirect_dest',
-        'ssl_cert',
-        'ssl_key',
-        'ssl_chain',
-        'use_hsts',
-        'disabled',
+    'fqdn',
+    'user',
+    'comment',
+    'ssl',
+    'socket_name',
+    'relative_docroot',
+    'redirect_dest',
 ))):
-    """A single <VirtualHost>.
+    """A single virtual host directive in one of the webserver configs.
 
-    The VirtualHost has a one-to-one mapping to <VirtualHost> directives and is
-    rendered by one invocation of the template.
+    A VirtualHost has a one-to-one mapping to <VirtualHost> (apache) or
+    server (nginx) directives and is rendered by one invocation of the
+    template. This means that some vhost lines might emit multiple
+    VirtualHost instances (redirect domains, etc.).
```

Comment:
I think that this is probably almost as good as it could be. It'd feel kind of awkward to subtype a whole class when only one component needs polymorphism. If this had to scale, I would define subtemplates so the main class and template don't have to know what kind of thing the vhost binds to and choose between Apache and Nginx configs by changing the jinja environment.